File: CS/WinWebSolution.Module/Controller.cs
using System;
using System.Collections.Generic;
using System.Text;
using DevExpress.ExpressApp;
using DevExpress.ExpressApp.SystemModule;

namespace WinWebSolution.Module {

    public class Mycontroller : ViewController<ListView> {
        NewObjectViewController controller1;
        ListViewProcessCurrentObjectController controller2;

        public Mycontroller() {
            TargetViewNesting = Nesting.Nested;
            TargetObjectType = typeof(Transaction);
        }

        protected override void OnActivated() {
            base.OnActivated();
            controller1 = Frame.GetController<NewObjectViewController>();
            if (controller1 != null) {
                controller1.ObjectCreated += new EventHandler<ObjectCreatedEventArgs>(controller1_ObjectCreated);
            }
            controller2 = Frame.GetController<ListViewProcessCurrentObjectController>();
            if (controller2 != null) {
                controller2.CustomProcessSelectedItem += new EventHandler<CustomProcessListViewSelectedItemEventArgs>(controller2_CustomProcessSelectedItem);
            }
        }

        protected override void OnDeactivated() {
            if (controller1 != null) {
                controller1.ObjectCreated -= new EventHandler<ObjectCreatedEventArgs>(controller1_ObjectCreated);
            }
            if (controller2 != null) {
                controller2.CustomProcessSelectedItem -= new EventHandler<CustomProcessListViewSelectedItemEventArgs>(controller2_CustomProcessSelectedItem);
            }
            base.OnDeactivated();
        }

        void controller1_ObjectCreated(object sender, ObjectCreatedEventArgs e) {
            Application.DetailViewCreating += new EventHandler<DetailViewCreatingEventArgs>(Application_DetailViewCreating);
        }

        void controller2_CustomProcessSelectedItem(object sender, CustomProcessListViewSelectedItemEventArgs e) {
            Application.DetailViewCreating += new EventHandler<DetailViewCreatingEventArgs>(Application_DetailViewCreating);
        }

        void Application_DetailViewCreating(object sender, DetailViewCreatingEventArgs e) {
            ((XafApplication)sender).DetailViewCreating -= new EventHandler<DetailViewCreatingEventArgs>(Application_DetailViewCreating);
            if (e.Obj is Transaction && Frame is NestedFrame) {
                Type parentType = ((NestedFrame)Frame).ViewItem.ObjectType;
                if (typeof(Consumer).IsAssignableFrom(parentType)) {
                    e.ViewID = "Transaction_DetailView_FromConsumer";
                }
                if (typeof(Producer).IsAssignableFrom(parentType)) {
                    e.ViewID = "Transaction_DetailView_FromProducer";
                }
            }
        }
    }
}

File: VB/WinWebSolution.Module/Controller.vb
Imports System
Imports System.Collections.Generic
Imports System.Text
Imports DevExpress.ExpressApp
Imports DevExpress.ExpressApp.SystemModule

Namespace WinWebSolution.Module

    Public Class Mycontroller
        Inherits ViewController(Of ListView)

        Private controller1 As NewObjectViewController
        Private controller2 As ListViewProcessCurrentObjectController

        Public Sub New()
            TargetViewNesting = Nesting.Nested
            TargetObjectType = GetType(Transaction)
        End Sub

        Protected Overrides Sub OnActivated()
            MyBase.OnActivated()
            controller1 = Frame.GetController(Of NewObjectViewController)()
            If controller1 IsNot Nothing Then
                AddHandler controller1.ObjectCreated, AddressOf controller1_ObjectCreated
            End If
            controller2 = Frame.GetController(Of ListViewProcessCurrentObjectController)()
            If controller2 IsNot Nothing Then
                AddHandler controller2.CustomProcessSelectedItem, AddressOf controller2_CustomProcessSelectedItem
            End If
        End Sub

        Protected Overrides Sub OnDeactivated()
            If controller1 IsNot Nothing Then
                RemoveHandler controller1.ObjectCreated, AddressOf controller1_ObjectCreated
            End If
            If controller2 IsNot Nothing Then
                RemoveHandler controller2.CustomProcessSelectedItem, AddressOf controller2_CustomProcessSelectedItem
            End If
            MyBase.OnDeactivated()
        End Sub

        Private Sub controller1_ObjectCreated(ByVal sender As Object, ByVal e As ObjectCreatedEventArgs)
            AddHandler Application.DetailViewCreating, AddressOf Application_DetailViewCreating
        End Sub

        Private Sub controller2_CustomProcessSelectedItem(ByVal sender As Object, ByVal e As CustomProcessListViewSelectedItemEventArgs)
            AddHandler Application.DetailViewCreating, AddressOf Application_DetailViewCreating
        End Sub

        Private Sub Application_DetailViewCreating(ByVal sender As Object, ByVal e As DetailViewCreatingEventArgs)
            RemoveHandler DirectCast(sender, XafApplication).DetailViewCreating, AddressOf Application_DetailViewCreating
            If TypeOf e.Obj Is Transaction AndAlso TypeOf Frame Is NestedFrame Then
                Dim parentType As Type = CType(Frame, NestedFrame).ViewItem.ObjectType
                If GetType(Consumer).IsAssignableFrom(parentType) Then
                    e.ViewID = "Transaction_DetailView_FromConsumer"
                End If
                If GetType(Producer).IsAssignableFrom(parentType) Then
                    e.ViewID = "Transaction_DetailView_FromProducer"
                End If
            End If
        End Sub
    End Class
End Namespace
